### ISSUE TITLE
Bug: [ban-types] should not apply to type definitions

### ISSUE THREAD

**Issue by JoshMcCullough** [typescript-eslint/typescript-eslint#7048]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.0.4&sourceType=module&code=C4TwDgpgBAcg9sAYgVwHYGNgEs6oDwAqAfFALxQFQQAewEqAJgM5Qoba5QD8UqEAbhABOUAFwUA3EA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6AIwEMnajTUMkRNGgB7aJHBgAviGlA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA

### Repro Code

```typescript
type NotFunction<T> = T extends Function ? never : T;
```

### ESLint Config

```javascript
{
  "rules": {
    "@typescript-eslint/ban-types": "error"
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

Using `Function` within a type declaration should not trigger this rule because it is not declaring a variable as type `Function`.

### Actual Result

Using `Function` in a type declaration triggers this rule.

### Additional Info

For instance, I need to define a type that specifically disallows anything that looks like a function (any arguments, any return type). I had to turn this rule off, though I want it to apply in every other case.

Bad:
```ts
const x:  Function = ...;
```

Okay:
```ts
type NotFunction<T> = T extends Function ? never : T;
```

**Comment by Josh-Cena** [comment]
Why not `T extends (...args: never) => unknown`? That's the most idiomatic way to define a function top-type, doesn't get confused by classes, and works with callables that don't extend from `Function`.

**Comment by bradzacher** [comment]
> Why not T extends (...args: never) => unknown? That's the most idiomatic way to define a function top-type

Exactly right - that's the way TS defines them in the lib types:
https://github.com/microsoft/TypeScript/blob/996aeb6aaf8d6a04f5b9f29085f062bca98642b2/src/lib/es5.d.ts#L1605-L1623

> doesn't get confused by classes, and works with callables that don't extend from Function.

Exactly right - `Function` is unsafe because it matches classes as well - which are technically functions but aren't really treated as such in TS itself.

```ts
class C {}
type Result = NotFunction<C>;
//    ^? C
// NOT never!
```

For these reasons - I agree with Josh - there's still not a good reason to use banned types. We will not be making a change here.
